src/audio/asrc/asrc_farrow.c

```diff
@@ -0,0 +1,1398 @@
+
+	/* Reset the timer values, since we want start a new control cycle. */
+	src_obj->time_value = 0;
+	src_obj->time_value_pull = 0;
+
+	/* set updated status */
+	src_obj->is_updated = true;
+
+	return ASRC_EC_OK;
+}
+
+void write_to_ring_buffer16(struct asrc_farrow  *src_obj,
+			    int16_t **input_buffers,
+			    uint32_t index_input_frame)
+{
+	uint8_t index_channel;
+
+	/* update the buffer_write_position */
+	(src_obj->buffer_write_position)++;
+
+	/* since it's a ring buffer we need a wrap around */
+	if (src_obj->buffer_write_position >= src_obj->buffer_length)
+		src_obj->buffer_write_position -= (src_obj->buffer_length >> 1);
+
+	/* write data to each channel */
+	for (index_channel = 0; index_channel < src_obj->num_channels;
+	     index_channel++) {
+		/*
+		 * Since we want the filter function to load 64 bit of
+		 * buffer data in one cycle, this function writes each
+		 * input sample to the buffer twice, one with an
+		 * offset of half the buffer size.  This way we don't
+		 * need a wrap around while loading #filter_length of
+		 * buffered samples.  The upper and lower half of the
+		 * buffer are redundant.  If the memory tradeoff is
+		 * critical, the buffer can be reduced to half the
+		 * size but therefore increased filter operations have
+		 * to be expected.
+		 */
+		switch (src_obj->input_format) {
+		case ASRC_IOF_DEINTERLEAVED:
+			src_obj->ring_buffers16[index_channel]
+				[src_obj->buffer_write_position]
+				= input_buffers[index_channel]
+				[index_input_frame];
+			src_obj->ring_buffers16[index_channel]
+				[src_obj->buffer_write_position -
+				 (src_obj->buffer_length >> 1)]
+				= input_buffers[index_channel]
+				[index_input_frame];
+			break;
+		case ASRC_IOF_INTERLEAVED:
+			src_obj->ring_buffers16[index_channel]
+				[src_obj->buffer_write_position]
+				= input_buffers[index_channel]
+				[src_obj->num_channels * index_input_frame];
+			src_obj->ring_buffers16[index_channel]
+				[src_obj->buffer_write_position -
+				 (src_obj->buffer_length >> 1)]
+				= input_buffers[index_channel]
+				[src_obj->num_channels * index_input_frame];
+			break;
+		default:
+			trace_asrc_error("write_to_ring_buffer: Invalid data "
+					 "format! (select interleaved or "
+					 "deinterleaved)");
+			break;
+		}
+	}
+}
+
+void write_to_ring_buffer32(struct asrc_farrow  *src_obj,
+			    int32_t **input_buffers,
+			    uint32_t index_input_frame)
+{
+	uint8_t index_channel;
+
+	/* update the buffer_write_position */
+	(src_obj->buffer_write_position)++;
+
+	/* since it's a ring buffer we need a wrap around */
+	if (src_obj->buffer_write_position >= src_obj->buffer_length)
+		src_obj->buffer_write_position -= (src_obj->buffer_length >> 1);
+
+	/* write data to each channel */
+	for (index_channel = 0; index_channel < src_obj->num_channels;
+	     index_channel++) {
+		/*
+		 * Since we want the filter function to load 64 bit of
+		 * buffer data in one cycle, this function writes each
```

Comment:
ah this is why you need this magic 8 factor?

Reply:
No, this duplicated data FIR delay line trick avoids the index/pointer wrap in the middle of MAC computation. It speeds up generic C code a lot (with cost of doubled delay line RAM) but it's a bit questionable for HiFi3 that works quite smoothly with circular buffers. I may revisit this later. Now when looking at write_to_ring_buffer16/32() there's an obvious optimization to be done. The switch-case can be placed outside channels for loop. I'm doing it now.

Reply:
done!